site/blog/2025-07-22-ha-postgres-full-sync.mdx

```diff
@@ -0,0 +1,84 @@
+
+Multigres will bring the flexibility of a pluggable durability policy. There will be a few predefined ones, but if your needs are bespoke, you can just write an extension. Examples of durability policies include:
+
+- cross-availability zone (cross-az): a replica in at least one other az must have my data
+- cross-region: same as cross-az, but for regions
+- at least two azs: replicas in two distinct azs must have my data
+- majority quorum: traditional consensus rules like RAFT
+- etc.
+
+The advantages of policy based durability go beyond ease of use. Essentially, the policy does not need to depend on the number of nodes in the quorum. This flexibility allows you to deploy more nodes or additional zones without affecting the performance of the cluster.
+
+More about how this works will be covered in subsequent blog posts related to generalized consensus.
+
+## Existing replication pitfalls
+
+Postgres has many options on how you can configure replication. They broadly fall into two categories:
+
+1. Asynchronous replication (async)
+    * The primary commits the data and returns success to the caller.
+    * changes are then sent asynchronously to the standby replica(s).
+    * replica applies the changes.
+2. Synchronous replication (full-sync)
+    * primary flushes the commit information to the [WAL](https://www.postgresql.org/docs/current/wal-intro.html).
+    * changes are shipped to the standby replica(s).
+    * replica acknowledges the message: Sub-configurations here control the exact time when the replica acknowledges the message and applies the changes.
+    * primary externalizes the transaction by releasing locks, etc., and returns success to the caller.
+
+The problem with async replication is that there are failure modes where you may lose data. For example, if a primary node crashes after acknowledging a commit but before sending the data to the replica, that commit is lost. In the case of a network partition, this can go on for a long time risking substantial data loss.
+
+Synchronous replication addresses this problem, but it also has its own pitfalls explained below. This [Video from Kukushkin](https://youtu.be/PFn9qRGzTMc?si=neDR5sgti16Hfni0) covers many possible ways full-sync replication can fail on you. Overall, the problems are due to one of the following reasons:
+
+1. The Primary writes to the WAL and crashes before sending the data to the standby:
+    
+    <img src="/img/blog/fsync1.svg" alt="fsync1" />
+    
+    * setups where there are clients subscribed to the WAL on the primary (like a CDC process), they would view and transmit this as a committed transaction. However, a failover will abandon this transaction, while the subscriber might have irreversibly acted on this.
+    * the primary is restarted after a failover, it will have a timeline that is incompatible with the rest of the cluster, and will not be able to rejoin the cluster.
```

Comment:
hmm, if it has older timeline, it can rejoin as replica, right?
and if it has more commits (that were not acknowledged by new primary), we can pg_rewind it, but it's data loss again..